subflow.go

```diff
@@ -0,0 +1,248 @@
+package multipath
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/getlantern/ema"
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+type pendingAck struct {
+	sz     int
+	sentAt time.Time
+}
+type subflow struct {
+	to   string
+	conn net.Conn
+	mpc  *mpConn
+
+	chClose       chan struct{}
+	closeOnce     sync.Once
+	sendQueue     chan *sendFrame
+	pendingAcks   map[uint64]pendingAck
+	muPendingAcks sync.Mutex
+	probeStart    atomic.Value // time.Time
+	emaRTT        *ema.EMA
+	tracker       StatsTracker
+}
+
+func startSubflow(to string, c net.Conn, mpc *mpConn, clientSide bool, probeStart time.Time, tracker StatsTracker) *subflow {
+	sf := &subflow{
+		to:          to,
+		conn:        c,
+		mpc:         mpc,
+		chClose:     make(chan struct{}),
+		sendQueue:   make(chan *sendFrame),
+		pendingAcks: make(map[uint64]pendingAck),
+		emaRTT:      ema.NewDuration(longRTT, rttAlpha),
+		tracker:     tracker,
+	}
+	go sf.sendLoop()
+	if clientSide {
+		initialRTT := time.Since(probeStart)
+		tracker.UpdateRTT(initialRTT)
+		sf.emaRTT.SetDuration(initialRTT)
+		// pong immediately so the server can calculate the RTT between when it
+		// sends the leading bytes and receives the pong frame.
+		sf.ack(frameTypePong)
+	} else {
+		sf.probeStart.Store(probeStart)
+	}
+	go func() {
+		if err := sf.readLoop(); err != nil {
+			log.Debugf("read loop to %s ended: %v", sf.to, err)
+		}
+	}()
+	return sf
+}
+
+func (sf *subflow) readLoop() (err error) {
+	ch := make(chan *frame)
+	r := byteReader{Reader: sf.conn}
+	go func() {
+		defer close(ch)
+		for {
+			sz, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			fn, err := ReadVarInt(r)
+			if err != nil {
+				sf.close()
+				return
+			}
+			if sz == 0 {
+				sf.gotACK(fn)
+				continue
+			}
+			log.Tracef("got frame %d from %s with %d bytes", fn, sf.to, sz)
+			buf := pool.Get(int(sz))
+			n, err := io.ReadFull(r, buf)
+			if err != nil {
+				pool.Put(buf)
+				sf.close()
+				return
+			}
+			if n != int(sz) {
+				panic(fmt.Sprintf("expect to read %d bytes, read %d", sz, n))
+			}
+			sf.ack(fn)
+			ch <- &frame{fn: fn, bytes: buf}
+			sf.tracker.OnRecv(sz)
+			select {
+			case <-sf.chClose:
+				return
+			default:
+				// continue
+			}
+		}
+	}()
+	probeTimer := time.NewTimer(randomize(probeInterval))
+	for {
+		select {
+		case frame := <-ch:
+			if frame == nil {
+				return
+			}
+			sf.mpc.recvQueue.add(frame)
+			if !probeTimer.Stop() {
+				<-probeTimer.C
+			}
+			probeTimer.Reset(randomize(probeInterval))
+		case <-probeTimer.C:
+			sf.probe()
+			probeTimer.Reset(randomize(probeInterval))
+		}
+	}
+}
+
+func (sf *subflow) sendLoop() {
+	for {
+		select {
+		case <-sf.chClose:
+			return
+		case frame := <-sf.sendQueue:
+			n, err := sf.conn.Write(frame.buf)
+			if err != nil {
+				sf.close()
```

Comment:
We may not always need to close the subflow, right?  For example, what if `netErr, ok := err.(net.Error); ok && netErr.Temporary()`?  I do see, however, that this is used to encourage the parent `mpConn` to try a different subflow. Maybe we could use some kind of mechanism to send this subflow to the back of the line.  This would be something of an optimization though, so I'd be fine with just leaving a comment or TODO or something.

Reply:
Good point. Added TODO and a line of log to see how often error happens.